### ISSUE TITLE
ValueError: check_hostname requires server_hostname when making GET requests

### ISSUE THREAD

**Issue by jhuang314** [kennethreitz/requests#2435]
I just built python 2.7.9 from source for RHEL5.7 and installed requests using easy_install.
I get the following error when running 

    r = requests.get('https://api.github.com/user', auth=('user', 'pass'))

```
Python 2.7.9 (default, Feb  2 2015, 17:03:23)
[GCC 4.1.2 20080704 (Red Hat 4.1.2-52)] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import requests
>>> r = requests.get('https://api.github.com/user', auth=('user', 'pass'))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/opt/python2.7/lib/python2.7/site-packages/requests-2.5.1-py2.7.egg/requests/api.py", line 65, in get
    return request('get', url, **kwargs)
  File "/opt/python2.7/lib/python2.7/site-packages/requests-2.5.1-py2.7.egg/requests/api.py", line 49, in request
    response = session.request(method=method, url=url, **kwargs)
  File "/opt/python2.7/lib/python2.7/site-packages/requests-2.5.1-py2.7.egg/requests/sessions.py", line 461, in request
    resp = self.send(prep, **send_kwargs)
  File "/opt/python2.7/lib/python2.7/site-packages/requests-2.5.1-py2.7.egg/requests/sessions.py", line 573, in send
    r = adapter.send(request, **kwargs)
  File "/opt/python2.7/lib/python2.7/site-packages/requests-2.5.1-py2.7.egg/requests/adapters.py", line 370, in send
    timeout=timeout
  File "/opt/python2.7/lib/python2.7/site-packages/requests-2.5.1-py2.7.egg/requests/packages/urllib3/connectionpool.py", line 518, in urlopen
    body=body, headers=headers)
  File "/opt/python2.7/lib/python2.7/site-packages/requests-2.5.1-py2.7.egg/requests/packages/urllib3/connectionpool.py", line 322, in _make_request
    self._validate_conn(conn)
  File "/opt/python2.7/lib/python2.7/site-packages/requests-2.5.1-py2.7.egg/requests/packages/urllib3/connectionpool.py", line 727, in _validate_conn
    conn.connect()
  File "/opt/python2.7/lib/python2.7/site-packages/requests-2.5.1-py2.7.egg/requests/packages/urllib3/connection.py", line 238, in connect
    ssl_version=resolved_ssl_version)
  File "/opt/python2.7/lib/python2.7/site-packages/requests-2.5.1-py2.7.egg/requests/packages/urllib3/util/ssl_.py", line 254, in ssl_wrap_socket
    return context.wrap_socket(sock)
  File "/opt/python2.7/lib/python2.7/ssl.py", line 350, in wrap_socket
    _context=self)
  File "/opt/python2.7/lib/python2.7/ssl.py", line 537, in __init__
    raise ValueError("check_hostname requires server_hostname")
ValueError: check_hostname requires server_hostname
>>>
```

**Comment by Lukasa** [comment]
Ok, so this happens in one of two cases: either `server_hostname` is `None`, or `ssl.HAS_SNI` is `False`. Can you print the result of `import ssl; print ssl.HAS_SNI` please?

**Comment by jhuang314** [comment]
I get `False`

**Comment by Lukasa** [comment]
Right, so the problem is that in Python 2.7.9 the assumption we make about `HAS_SNI` and hostname verification is wrong. This is actually a bug in [urllib3](https://github.com/shazow/urllib3), so we need to open it over there.

**Comment by gowlin** [comment]
Forgive my ignorance, I'm trying to follow what's going on with this issue and the linked issues since I'm experiencing this problem right now.  Is there a workaround in the meantime before this change makes it into released versions?  Currently I have requests-2.5.0 installed with pip and virtualenv, which seems to pull in its own copy of urllib3.  Is there a different compatible version of requests I can install to workaround?

**Comment by Trim21** [comment]
It's broken again with `reqeusts==2.25.1` and `urllib3=1.26.4'

```console
$ cat a.py
import sys
import ssl

import requests

print(sys.version, sys.version_info)
print(ssl.HAS_SNI)
print(requests.get("https://httpbin.org/anything").text)
$ python --version
Python 3.7.9
$ python a.py
3.7.9 (tags/v3.7.9:13c94747c7, Aug 17 2020, 18:58:18) [MSC v.1900 64 bit (AMD64)] sys.version_info(major=3, minor=7, micro=9, releaselevel='final', serial=0)
True
Traceback (most recent call last):
  File "a.py", line 8, in <module>
    print(requests.get("https://httpbin.org/anything").text)
  File "C:\Users\Trim21\.venv\BGmi\lib\site-packages\requests\api.py", line 76, in get
    return request('get', url, params=params, **kwargs)
  File "C:\Users\Trim21\.venv\BGmi\lib\site-packages\requests\api.py", line 61, in request
    return session.request(method=method, url=url, **kwargs)
  File "C:\Users\Trim21\.venv\BGmi\lib\site-packages\requests\sessions.py", line 542, in request
    resp = self.send(prep, **send_kwargs)
  File "C:\Users\Trim21\.venv\BGmi\lib\site-packages\requests\sessions.py", line 655, in send
    r = adapter.send(request, **kwargs)
  File "C:\Users\Trim21\.venv\BGmi\lib\site-packages\requests\adapters.py", line 449, in send
    timeout=timeout
  File "C:\Users\Trim21\.venv\BGmi\lib\site-packages\urllib3\connectionpool.py", line 696, in urlopen
    self._prepare_proxy(conn)
  File "C:\Users\Trim21\.venv\BGmi\lib\site-packages\urllib3\connectionpool.py", line 964, in _prepare_proxy
    conn.connect()
  File "C:\Users\Trim21\.venv\BGmi\lib\site-packages\urllib3\connection.py", line 359, in connect
    conn = self._connect_tls_proxy(hostname, conn)
  File "C:\Users\Trim21\.venv\BGmi\lib\site-packages\urllib3\connection.py", line 506, in _connect_tls_proxy
    ssl_context=ssl_context,
  File "C:\Users\Trim21\.venv\BGmi\lib\site-packages\urllib3\util\ssl_.py", line 432, in ssl_wrap_socket
    ssl_sock = _ssl_wrap_socket_impl(sock, context, tls_in_tls)
  File "C:\Users\Trim21\.venv\BGmi\lib\site-packages\urllib3\util\ssl_.py", line 474, in _ssl_wrap_socket_impl
    return ssl_context.wrap_socket(sock)
  File "C:\Users\Trim21\scoop\apps\python37\current\lib\ssl.py", line 423, in wrap_socket
    session=session
  File "C:\Users\Trim21\scoop\apps\python37\current\lib\ssl.py", line 827, in _create
    raise ValueError("check_hostname requires server_hostname")
ValueError: check_hostname requires server_hostname
```
